inject/src/main/java/io/micronaut/inject/ast/utils/AstBeanPropertiesUtils.java

```diff
@@ -0,0 +1,384 @@
+            return false;
+        }
+        if (accessKinds.contains(BeanProperties.AccessKind.FIELD)) {
+            return isAccessible(fieldElement, visibility);
+        }
+        return false;
+    }
+
+    private static boolean canMethodBeUsedForAccess(MethodElement methodElement,
+                                                    Set<BeanProperties.AccessKind> accessKinds,
+                                                    BeanProperties.Visibility visibility) {
+        return accessKinds.contains(BeanProperties.AccessKind.METHOD) && isAccessible(methodElement, visibility);
+    }
+
+    private static boolean isAccessible(MemberElement memberElement, BeanProperties.Visibility visibility) {
+        switch (visibility) {
+            case DEFAULT:
+                return !memberElement.isPrivate() && (memberElement.isAccessible() || memberElement.getDeclaringType().hasDeclaredStereotype(BeanProperties.class));
+            case PUBLIC:
+                return memberElement.isPublic();
+            default:
+                return false;
+        }
+    }
+
+    public static List<MethodElement> getSubtypeFirstMethods(ClassElement classElement) {
+        List<MethodElement> methods = classElement.getEnclosedElements(ElementQuery.ALL_METHODS.onlyInstance());
+        List<MethodElement> result = new ArrayList<>(methods.size());
+        List<MethodElement> other = new ArrayList<>(methods.size());
+        // Process subtype methods first
+        for (MethodElement methodElement : methods) {
+            if (methodElement.getDeclaringType().equals(classElement)) {
+                other.add(methodElement);
+            } else {
+                result.add(methodElement);
+            }
+        }
+        result.addAll(other);
+        return result;
+
+    }
+
+    public static List<FieldElement> getSubtypeFirstFields(ClassElement classElement) {
+        List<FieldElement> fields = classElement.getEnclosedElements(ElementQuery.ALL_FIELDS);
+        List<FieldElement> result = new ArrayList<>(fields.size());
+        List<FieldElement> other = new ArrayList<>(fields.size());
+        // Process subtype fields first
+        for (FieldElement fieldElement : fields) {
+            if (fieldElement.getDeclaringType().equals(classElement)) {
+                other.add(fieldElement);
+            } else {
+                result.add(fieldElement);
+            }
+        }
+        result.addAll(other);
+        return result;
+    }
+
+    private static boolean shouldExclude(Set<String> includes, Set<String> excludes, String propertyName) {
+        if (!includes.isEmpty() && !includes.contains(propertyName)) {
+            return true;
+        }
+        return !excludes.isEmpty() && excludes.contains(propertyName);
+    }
+
+    /**
+     * Internal holder class for getters and setters.
+     */
+    @SuppressWarnings("VisibilityModifier")
+    public static class BeanPropertyData {
```

Comment:
declare `final`

Reply:
does this need to be public?

Reply:
Yes, it's used to build the property instance.